### ISSUE TITLE
Bug: RuleTester can't parse fixture files unless they are nested under two fixture folders

### ISSUE THREAD

**Issue by justinfagnani** [typescript-eslint/typescript-eslint#8271]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

I expected that the RuleTester would be able to parse my fixture files, when I followed [the type-aware test set up instructions](https://typescript-eslint.io/packages/rule-tester/#type-aware-testing).

Instead I'm getting this error:

```
  AssertionError [ERR_ASSERTION]: A fatal parsing error occurred: Parsing error: ESLint was configured to run on `<tsconfigRootDir>/fixture/file.ts` using `parserOptions.project`: <tsconfigRootDir>/tsconfig.json
  However, that TSConfig does not include this file. Either:
  - Change ESLint's list of included files to not include this file
  - Change that TSConfig to include this file
  - Create a new TSConfig that includes this file and include it in your parserOptions.project
  See the typescript-eslint docs for more info: https://typescript-eslint.io/linting/troubleshooting#i-get-errors-telling-me-eslint-was-configured-to-run--however-that-tsconfig-does-not--none-of-those-tsconfigs-include-this-file
      at RuleTester.runRuleForItem (/Users/justinfagnani/Projects/Lit/tmp/ts-eslint-starter/node_modules/@typescript-eslint/rule-tester/src/RuleTester.ts:590:11)
      at RuleTester._RuleTester_testInvalidTemplate (/Users/justinfagnani/Projects/Lit/tmp/ts-eslint-starter/node_modules/@typescript-eslint/rule-tester/src/RuleTester.ts:706:25)
      at TestContext.<anonymous> (/Users/justinfagnani/Projects/Lit/tmp/ts-eslint-starter/node_modules/@typescript-eslint/rule-tester/src/RuleTester.ts:426:40)
      at Test.runInAsyncScope (node:async_hooks:206:9)
      at Test.run (node:internal/test_runner/test:631:25)
      at Test.start (node:internal/test_runner/test:542:17)
      at node:internal/test_runner/test:947:71
      at node:internal/per_context/primordials:488:82
      at new Promise (<anonymous>)
      at new SafePromise (node:internal/per_context/primordials:456:29) {
```

Moving the fixture files (`file.ts` and `react.tsx`) to a _nested_ fixture folder removes the parsing error, however then there is a problem where the empty file contents are not replaced by the code in the tests.

The nested structure that at least parses is:

```
fixture
├── fixture
│   ├── file.ts
│   └── react.tsx
└── tsconfig.json
```

### Reproduction Repository Link

https://github.com/justinfagnani/typescript-eslint-issue

### Repro Steps

```sh
$ git clone https://github.com/justinfagnani/typescript-eslint-issue
$ cd typescript-eslint-issue
$ npm ci
$ npm test
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/parser`            | `6.19.0` |
| `@typescript-eslint/rule-tester`       | `6.19.0` |
| `@typescript-eslint/utils`             | `6.19.0` |
| `typescript`                           | `5.3.3` |
| `eslint`                               | `8.56.0` |
| `node`                                 | `21.6.0` |

**Comment by bradzacher** [comment]
This looks like a bad interaction between `rule-tester` and `typescript-estree` when you use a relative path for `tsconfigRootDir`. We don't encounter this internally because we use an absolute path.

Essentially `rule-tester` makes the filename relative to the tsconfig dir:
https://github.com/typescript-eslint/typescript-eslint/blob/01556f50376b87eff21c42d121b627e3a6f003f0/packages/rule-tester/src/RuleTester.ts#L199-L205

EG `file.ts` => `./fixtures/file.ts`

Then `typescript-estree` enforces that it's an absolute path using the only disk location we have - `tsconfigRootDir`:
https://github.com/typescript-eslint/typescript-eslint/blob/01556f50376b87eff21c42d121b627e3a6f003f0/packages/typescript-estree/src/parseSettings/createParseSettings.ts#L99-L104

This doesn't happen in normal ESLint runs because ESLint passes through an absolute path:
https://github.com/eslint/eslint/blob/38b9b06695f88c70441dd15ae5d97ffd8088be23/lib/eslint/eslint.js#L997

The fix here probably should be to make `rule-tester` absolutise the path so it matches ESLint.

**Comment by justinfagnani** [comment]
Awesome, thanks for figuring that out @bradzacher!

As you say, everything works if I set up the rule tester with an absolute path:
```ts
const fixtureRootDir = new URL("../../fixture", import.meta.url).pathname;

export const ruleTester = new RuleTester({
  parser: "@typescript-eslint/parser",
  parserOptions: {
    tsconfigRootDir: fixtureRootDir,
    project: "./tsconfig.json",
  },
});
```

If I can find some time soon, I'll se if I can submit a PR to fix this too.

What should relate paths be relative to? The project or cwd? It's not possible to make it relative to the defining module without the module passing in `import.meta.url`. Another option is to require absolute paths and make the user handle it.

**Comment by bradzacher** [comment]
We should mirror eslint - so it should be considered relative to the CWD (https://github.com/eslint/eslint/blob/38b9b06695f88c70441dd15ae5d97ffd8088be23/lib%2Feslint%2Feslint.js)

**Comment by sakgoyal** [comment]
how do I fix this? do I just wait for an update to eslint?

**Comment by JoshuaKGoldberg** [comment]
> how do I fix this

![Screenshot of labels on this issue: 'accepting prs', 'bug', 'package: rule-tester'](https://github.com/typescript-eslint/typescript-eslint/assets/3335181/b235e02f-4bcf-441a-ac17-800ecc6b221a)

https://typescript-eslint.io/contributing

😛

Otherwise, see https://github.com/typescript-eslint/typescript-eslint/issues/8271#issuecomment-1899554557.
